drivers/xen/gnttab.c

```diff
@@ -34,53 +34,60 @@
 LOG_MODULE_REGISTER(xen_gnttab);
 
 /* Timeout for grant table ops retrying */
-#define GOP_RETRY_DELAY 200
+#define GOP_RETRY_DELAY		200
+#define DT_GNTTAB_SIZE		DT_REG_SIZE_BY_IDX(DT_INST(0, xen_xen), 0)
+#define GNT_ENTRIES_PER_FRAME	(XEN_PAGE_SIZE / sizeof(grant_entry_v1_t))
+#define GNTTAB_LAST_GREF	UINT32_MAX
 
-#define GNTTAB_SIZE DT_REG_SIZE_BY_IDX(DT_INST(0, xen_xen), 0)
-BUILD_ASSERT(!(GNTTAB_SIZE % XEN_PAGE_SIZE), "Size of gnttab have to be aligned on XEN_PAGE_SIZE");
+BUILD_ASSERT(!(DT_GNTTAB_SIZE % XEN_PAGE_SIZE), "Size of gnttab have to be aligned on XEN_PAGE_SIZE");
+BUILD_ASSERT(DT_GNTTAB_SIZE <= CONFIG_KERNEL_VM_SIZE);
 
-/* NR_GRANT_FRAMES must be less than or equal to that configured in Xen */
-#define NR_GRANT_FRAMES (GNTTAB_SIZE / XEN_PAGE_SIZE)
-#define NR_GRANT_ENTRIES \
-	(NR_GRANT_FRAMES * XEN_PAGE_SIZE / sizeof(grant_entry_v1_t))
-
-BUILD_ASSERT(GNTTAB_SIZE <= CONFIG_KERNEL_VM_SIZE);
 DEVICE_MMIO_TOPLEVEL_STATIC(grant_tables, DT_INST(0, xen_xen));
 
 static struct gnttab {
-	struct k_sem sem;
+	struct k_mutex lock;
+	struct k_condvar condvar;
+	unsigned long nr_grant_frames;
+	unsigned long max_grant_frames;
 	grant_entry_v1_t *table;
-	grant_ref_t gref_list[NR_GRANT_ENTRIES];
+	grant_ref_t *gref_list;
 } gnttab;
 
-static grant_ref_t get_free_entry(void)
+static int extend_gnttab(void);
+
+static grant_ref_t get_grant_entry(void)
 {
+	int rc;
 	grant_ref_t gref;
-	unsigned int flags;
 
-	k_sem_take(&gnttab.sem, K_FOREVER);
+	k_mutex_lock(&gnttab.lock, K_FOREVER);
+	if(gnttab.gref_list[0] == GNTTAB_LAST_GREF) {
+		/* Map one more frame if possible, need to hold mutex */
+		rc = extend_gnttab();
+		if (rc) {
+			/* Failed to extend grant table, so need to wait for free entry */
```

Comment:
Do we really need to wait for a free entry? Why we can't just return an error here? 

Your approach may block some device driver for undefined amount of time. Which is pretty bad, but is double bad for an RTOS.

Reply:
That was one of the controversial point in here. Previous implementation would wait on semaphore, but it was not dynamically expandable. So, now I think we can remove condvar and just return error (made in last version).